src/Servers/HttpSys/src/MessagePump.cs

```diff
@@ -275,4 +225,87 @@ public void Dispose()
 
         Listener.Dispose();
     }
+
+    private sealed class AcceptLoop : IThreadPoolWorkItem
+    {
+        private readonly AsyncAcceptContext _asyncAcceptContext;
+        private readonly MessagePump _messagePump;
+        private readonly bool _preferInlineScheduling;
+
+        public AcceptLoop(AsyncAcceptContext asyncAcceptContext,
+                          MessagePump messagePump)
+        {
+            _asyncAcceptContext = asyncAcceptContext;
+            _messagePump = messagePump;
+            _preferInlineScheduling = _messagePump._options.UnsafePreferInlineScheduling;
+        }
+
+        public void Execute()
+        {
+            _ = ExecuteAsync();
+        }
+
+        private async Task ExecuteAsync()
+        {
+            while (!_messagePump.Stopping)
+            {
+                // Receive a request
+                RequestContext requestContext;
+                try
+                {
+                    requestContext = await _messagePump.Listener.AcceptAsync(_asyncAcceptContext);
+
+                    if (!_messagePump.Listener.ValidateRequest(requestContext))
+                    {
+                        // Dispose the request
+                        requestContext.ReleasePins();
+                        requestContext.Dispose();
+
+                        // If either of these is false then a response has already been sent to the client, so we can accept the next request
+                        continue;
+                    }
+                }
+                catch (Exception exception)
+                {
+                    Debug.Assert(_messagePump.Stopping);
+                    if (_messagePump.Stopping)
+                    {
+                        Log.AcceptErrorStopping(_messagePump._logger, exception);
+                    }
+                    else
+                    {
+                        Log.AcceptError(_messagePump._logger, exception);
```

Comment:
uber nit: exception vs ex is inconsistent in catches
```suggestion
                catch (Exception ex)
                {
                    Debug.Assert(_messagePump.Stopping);
                    if (_messagePump.Stopping)
                    {
                        Log.AcceptErrorStopping(_messagePump._logger, ex);
                    }
                    else
                    {
                        Log.AcceptError(_messagePump._logger, ex);
```